internal/alb/tg/targethealth.go

```diff
@@ -0,0 +1,339 @@
+)
+
+type targetGroupWatch struct {
+	// ingress is the ingress for the target group
+	ingress *extensions.Ingress
+
+	// backend is the ingress backend for the target group
+	backend *extensions.IngressBackend
+
+	// podsToReconcile is the list of pods whose conditions have not been reconciled yet
+	targetsToReconcile []*elbv2.TargetDescription
+
+	cancel context.CancelFunc
+	mux    sync.Mutex
+}
+type targetGroupWatches map[string]*targetGroupWatch
+
+// TargetHealthController provides functionality to reconcile pod condition status from target health of targets in a target group
+type TargetHealthController interface {
+	// Reconcile ensures the target group targets in AWS matches the targets configured in the ingress backend.
+	SyncTargetsForReconciliation(ctx context.Context, t *Targets, desiredTargets []*elbv2.TargetDescription) error
+	RemovePodConditions(ctx context.Context, t *Targets, targets []*elbv2.TargetDescription) error
+	StopReconcilingPodConditionStatus(tgArn string)
+}
+
+// NewTargetHealthController constructs a new target health controller
+func NewTargetHealthController(cloud aws.CloudAPI, store store.Storer, endpointResolver backend.EndpointResolver, client client.Client) TargetHealthController {
+	return &targetHealthController{
+		cloud:            cloud,
+		store:            store,
+		endpointResolver: endpointResolver,
+		client:           client,
+		tgWatches:        make(targetGroupWatches),
+	}
+}
+
+type targetHealthController struct {
+	cloud            aws.CloudAPI
+	store            store.Storer
+	endpointResolver backend.EndpointResolver
+	client           client.Client
+	tgWatches        targetGroupWatches
+	tgWatchesMux     sync.Mutex
+}
+
+// SyncTargetsForReconciliation starts a go routine for reconciling pod condition statuses for the given targets in the background until they are healthy in the target group
+func (c *targetHealthController) SyncTargetsForReconciliation(ctx context.Context, t *Targets, desiredTargets []*elbv2.TargetDescription) error {
+	conditionType := podConditionTypeForIngressBackend(t.Ingress, t.Backend)
+
+	targetsToReconcile, err := c.filterTargetsNeedingReconciliation(conditionType, t, desiredTargets)
+	if err != nil {
+		return err
+	}
+
+	// create, update or remove targetGroupWatch for this target group;
+	// a targetGroupWatch exists as long as there are targets in the target group whose pod condition statuses need to be reconciled;
+	// while the targetGroupWatch exists, a go routine regularly monitors the target health of the targets in the target group and updates the pod condition status for the corresponding pods
+	c.tgWatchesMux.Lock()
+	defer c.tgWatchesMux.Unlock()
+
+	tgWatch, ok := c.tgWatches[t.TgArn]
+	if ok {
+		// targetGroupWatch for this target group already exists
+		if len(targetsToReconcile) == 0 {
+			tgWatch.cancel()
+			delete(c.tgWatches, t.TgArn)
+		} else {
+			tgWatch.mux.Lock()
+			tgWatch.ingress = t.Ingress
```

Comment:
what's the purpose here. why only ingress & targets been updated but not backend?

Reply:
That's actually a left-over. The ingress update was so that changes to the target health-specific annotations would be detected while the go routine was running, but since these annotations are gone now, this is not needed anymore. (I guess updates to the annotations for health check interval / health threshold count shouldn't matter.)

Reply:
then we may consider make tgWatch depends on `healthCheckInterval` instead of both Ingress & Backend.

Reply:
Not sure what you mean, it needs both ingress and backend, they just don't need to be updated while the go routine is running.

Reply:
i didn't notice that ingress & backend is used in logs, we can keep current behavior and skip the update

Reply:
It's also used for retrieving the pods corresponding to the targets.